pkg/vdri/trustbloc/didconfiguration/helpers.go

```diff
@@ -68,3 +72,141 @@ func createDomainLinkageAssertion(
 		JWT: jwsCompact,
 	}, nil
 }
+
+// VerifyDIDConfiguration verifies a DID configuration, using the given VDRI to resolve the DID.
+//   returns a list of the DIDs that were successfully authenticated to this domain
+func VerifyDIDConfiguration(domain string, configuration *models.DIDConfiguration, doc *did.Doc) ([]string, error) {
+	didSet := map[string]struct{}{}
+
+	var errs []string
+
+	for i, dla := range configuration.Entries {
+		err := ValidateDomainLinkageAssertion(domain, dla, doc)
+		if err != nil {
+			log.Debugf("domain linkage assertion %v for %s invalid", i, domain)
+
+			errs = append(errs, err.Error())
+		} else {
+			didSet[dla.DID] = struct{}{}
+		}
+	}
+
+	dids := make([]string, 0, len(didSet))
+	for id := range didSet {
+		dids = append(dids, id)
+	}
+
+	if len(dids) == 0 {
+		errMsg := ""
+		for _, ems := range errs {
+			errMsg += "`" + ems + "`, "
+		}
+
+		return nil, fmt.Errorf("all domain linkage assertions invalid for domain %s: %s", domain, errMsg)
+	}
+
+	return dids, nil
+}
+
+// ValidateDomainLinkageAssertion validates a domain linkage assertion, using the given VDRI to resolve the DID
+func ValidateDomainLinkageAssertion(domain string, assertion models.DomainLinkageAssertion, doc *did.Doc) error {
+	jws, err := jose.ParseSigned(assertion.JWT)
+	if err != nil {
+		return fmt.Errorf("cannot parse assertion JWT: %w", err)
+	}
+
+	var claims models.DomainLinkageAssertionClaims
+
+	rawClaims := jws.UnsafePayloadWithoutVerification()
+	err = json.Unmarshal(rawClaims, &claims)
+
+	if err != nil {
+		return fmt.Errorf("cannot parse assertion JWT claims: %w", err)
+	}
+
+	if claims.ISS != assertion.DID {
+		return fmt.Errorf("assertion DID does not match signature")
+	}
+
+	if claims.Domain != domain {
+		return fmt.Errorf("assertion domain does not match host domain")
+	}
+
+	if claims.Exp != 0 && claims.Exp <= time.Now().Unix() {
+		return fmt.Errorf("assertion has expired")
+	}
+
+	_, err = VerifyDIDSignature(jws, doc)
+
+	return err
+}
+
+// VerifyDIDSignature verify a signature using a DID doc
+func VerifyDIDSignature(jws *jose.JSONWebSignature, doc *did.Doc) ([]byte, error) {
+	if jws == nil {
+		return nil, fmt.Errorf("jws is nil")
+	}
+
+	jwkList := getJWKs(doc)
+
+	errs := ""
+
+	var val []byte
+
+	var err error
+
+	verified := false
+
+	for _, key := range jwkList {
+		if key == nil || key.Key == nil {
+			return nil, fmt.Errorf("key is nil")
+		}
+
+		val, err = jws.Verify(key.Key)
+		if err == nil {
+			verified = true
+			break
+		} else {
+			errs += err.Error() + ", "
+		}
+	}
+
+	if !verified {
+		docMsg := ""
+
+		docBytes, err := doc.JSONBytes()
+		if err == nil {
+			docMsg = " using doc:\n" + string(docBytes)
+		}
+
+		return nil, fmt.Errorf("failed to verify: %s%s", errs, docMsg)
+	}
+
+	return val, nil
+}
+
+func getJWKs(doc *did.Doc) []*jose2.JWK {
+	var jwkList []*jose2.JWK
+
+	for _, pk := range doc.PublicKey {
+		jwk := pk.JSONWebKey()
+		if jwk == nil {
+			// TODO log?
```

Comment:
Are you supposed to verify signature against all public keys with JWK in the document or against specific key in the document? I am not sure if you need TODO: here.

Reply:
Verify against all - we could discuss if we should instead identify a particular key for each type of signature (eg, a key for did-configuration signing/verification, a key for did-method config signing/verification), I can see how that could be a good idea.